diff --git a/net/ipv4/fib_semantics.c b/net/ipv4/fib_semantics.c
index 91a3d6d33..765371fcf 100644
--- a/net/ipv4/fib_semantics.c
+++ b/net/ipv4/fib_semantics.c
@@ -47,6 +47,7 @@
 #include <net/fib_notifier.h>
 #include <net/addrconf.h>
 
+#include <linux/vpnhide.h>
 #include "fib_lookup.h"
 
 static DEFINE_SPINLOCK(fib_info_lock);
@@ -1803,6 +1804,9 @@ int fib_dump_info(struct sk_buff *skb, u32 portid, u32 seq, int event,
 	struct nlmsghdr *nlh;
 	struct rtmsg *rtm;
 
+	if (vpnhide_hide_fib_dump(fri))
+		return 0;
+
 	nlh = nlmsg_put(skb, portid, seq, event, sizeof(*rtm), flags);
 	if (!nlh)
 		return -EMSGSIZE;
