diff --git a/fs/readdir.c b/fs/readdir.c
index e6f4c7b88..34181fcdb 100644
--- a/fs/readdir.c
+++ b/fs/readdir.c
@@ -22,6 +22,7 @@
 #include <linux/compat.h>
 
 #include <linux/uaccess.h>
+#include <linux/vpnhide.h>
 
 int iterate_dir(struct file *file, struct dir_context *ctx)
 {
@@ -47,11 +48,14 @@ int iterate_dir(struct file *file, struct dir_context *ctx)
 
 	res = -ENOENT;
 	if (!IS_DEADDIR(inode)) {
+		bool vh_dir = vpnhide_readdir_begin(file, ctx);
 		ctx->pos = file->f_pos;
 		if (shared)
 			res = file->f_op->iterate_shared(file, ctx);
 		else
 			res = file->f_op->iterate(file, ctx);
+		if (vh_dir)
+			vpnhide_readdir_end(ctx);
 		file->f_pos = ctx->pos;
 		fsnotify_access(file);
 		file_accessed(file);
