diff --git a/fs/stat.c b/fs/stat.c
index 0fda4b6b8..6ffb37a1a 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -20,6 +20,7 @@
 
 #include <linux/uaccess.h>
 #include <asm/unistd.h>
+#include <linux/vpnhide.h>
 
 /**
  * generic_fillattr - Fill in the basic attributes from the inode struct
@@ -113,7 +114,10 @@ int vfs_getattr(const struct path *path, struct kstat *stat,
 	retval = security_inode_getattr(path);
 	if (retval)
 		return retval;
-	return vfs_getattr_nosec(path, stat, request_mask, query_flags);
+	retval = vfs_getattr_nosec(path, stat, request_mask, query_flags);
+	if (!retval && vpnhide_should_hide_dentry(path->dentry))
+		retval = -ENOENT;
+	return retval;
 }
 EXPORT_SYMBOL(vfs_getattr);
 
