diff --git a/fs/stat.c b/fs/stat.c
index 068fdbcc9..51218e631 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -17,6 +17,7 @@
 
 #include <asm/uaccess.h>
 #include <asm/unistd.h>
+#include <linux/vpnhide.h>
 
 void generic_fillattr(struct inode *inode, struct kstat *stat)
 {
@@ -69,7 +70,10 @@ int vfs_getattr(struct path *path, struct kstat *stat)
 	retval = security_inode_getattr(path);
 	if (retval)
 		return retval;
-	return vfs_getattr_nosec(path, stat);
+	retval = vfs_getattr_nosec(path, stat);
+	if (!retval && vpnhide_should_hide_dentry(path->dentry))
+		retval = -ENOENT;
+	return retval;
 }
 
 EXPORT_SYMBOL(vfs_getattr);
