diff --git a/net/ipv6/route.c b/net/ipv6/route.c
index cf05b953f..d5814a06b 100644
--- a/net/ipv6/route.c
+++ b/net/ipv6/route.c
@@ -63,6 +63,7 @@
 #include <net/ip_tunnels.h>
 #include <net/l3mdev.h>
 #include <trace/events/fib6.h>
+#include <linux/vpnhide.h>
 
 #include <asm/uaccess.h>
 
@@ -3181,6 +3182,9 @@ static int rt6_fill_node(struct net *net,
 		}
 	}
 
+	if (vpnhide_hide_rt6(rt, NULL))
+		return 0;
+
 	nlh = nlmsg_put(skb, portid, seq, type, sizeof(*rtm), flags);
 	if (!nlh)
 		return -EMSGSIZE;
