diff --git a/net/ipv6/ip6_fib.c b/net/ipv6/ip6_fib.c
index 5ff67cb8b..19dc24501 100644
--- a/net/ipv6/ip6_fib.c
+++ b/net/ipv6/ip6_fib.c
@@ -37,6 +37,7 @@
 
 #include <net/ip6_fib.h>
 #include <net/ip6_route.h>
+#include <linux/vpnhide.h>
 
 static struct kmem_cache *fib6_node_kmem __read_mostly;
 
@@ -2285,6 +2286,11 @@ static int ipv6_route_seq_show(struct seq_file *seq, void *v)
 	struct ipv6_route_iter *iter = seq->private;
 	const struct net_device *dev;
 
+	if (vpnhide_hide_fib6_route(rt)) {
+		iter->w.leaf = NULL;
+		return 0;
+	}
+
 	seq_printf(seq, "%pi6 %02x ", &rt->fib6_dst.addr, rt->fib6_dst.plen);
 
 #ifdef CONFIG_IPV6_SUBTREES
