diff --git a/net/ipv4/fib_semantics.c b/net/ipv4/fib_semantics.c
index 9573cd242..c2f4c12d3 100644
--- a/net/ipv4/fib_semantics.c
+++ b/net/ipv4/fib_semantics.c
@@ -46,6 +46,7 @@
 #include <net/lwtunnel.h>
 #include <net/fib_notifier.h>
 
+#include <linux/vpnhide.h>
 #include "fib_lookup.h"
 
 static DEFINE_SPINLOCK(fib_info_lock);
@@ -1295,6 +1296,9 @@ int fib_dump_info(struct sk_buff *skb, u32 portid, u32 seq, int event,
 	struct nlmsghdr *nlh;
 	struct rtmsg *rtm;
 
+	if (vpnhide_hide_fib_dump_raw(fi, dst, dst_len))
+		return 0;
+
 	nlh = nlmsg_put(skb, portid, seq, event, sizeof(*rtm), flags);
 	if (!nlh)
 		return -EMSGSIZE;
