diff --git a/net/core/rtnetlink.c b/net/core/rtnetlink.c
index d7e45067b..d7fb95652 100644
--- a/net/core/rtnetlink.c
+++ b/net/core/rtnetlink.c
@@ -49,6 +49,7 @@
 #include <net/udp.h>
 #include <net/tcp.h>
 #include <net/sock.h>
+#include <linux/vpnhide.h>
 #include <net/pkt_sched.h>
 #include <net/fib_rules.h>
 #include <net/rtnetlink.h>
@@ -1719,6 +1720,10 @@ static int rtnl_fill_ifinfo(struct sk_buff *skb,
 	struct nlmsghdr *nlh;
 	struct Qdisc *qdisc;
 
+	if (dev &&
+	    vpnhide_should_hide_dev(dev, VPNHIDE_HID_RTNL_FILL_IFINFO))
+		return 0;
+
 	ASSERT_RTNL();
 	nlh = nlmsg_put(skb, pid, seq, type, sizeof(*ifm), flags);
 	if (nlh == NULL)
